### PR TITLE
Enforce exact kind-445 ingress shape

### LINKED ISSUE #1007: Enforce exact kind-445 tag shape without treating expiration as validity
## Summary

Make Nostr kind-445 group-message ingress enforce the adopted exact tag shape. Correct the prior issue scope: NIP-40 `expiration` is transport metadata and is **not** a Marmot event-validity rule. A client may receive/process an otherwise valid event after its expiration timestamp.

DNS/host rebinding controls are also a local outbound dial-safety concern, not kind-445 syntax; they are covered by the transport/dial work rather than this parser issue.

## Normative source

Adopted by marmot-protocol/marmot#408:

- `transports/nostr.md`
- `foundation/wire-envelopes.md`
- `protocol-core/inbound-processing.md`

## Start here

- `crates/transport-nostr-peeler/src/peeler.rs`
- `crates/transport-nostr-peeler/src/lib.rs`
- Nostr event building/parsing in `crates/transport-nostr-adapter/`
- relay runtime and peeler tests

## Required behavior

- Require the exact kind-445 tag names, counts, arities, marker values, and byte encodings specified for a Marmot group event.
- Reject extra or duplicate protocol tags where the spec requires an exact shape; do not accept a valid required subset surrounded by arbitrary tags.
- Validate the event signature/ID and outer-envelope fields before handing ciphertext to the MLS engine.
- Keep routing ID and raw MLS group ID distinct at every boundary.
- Treat a syntactically valid NIP-40 `expiration` tag only according to the adopted transport behavior. Do not reject solely because wall-clock time is after expiration.
- Produce typed malformed/unauthenticated/wrong-recipient/unknown-group outcomes without leaking identifiers into logs.

## Out of scope

- Relay marker/prior-route behavior (#1057).
- Durable outbound fanout (#1020).
- DNS resolution and host-safety policy except ensuring this parser does not pretend to provide it.

## Acceptance criteria

- Table-driven tests cover missing, duplicate, extra, misordered where relevant, wrong-arity, wrong-marker, malformed-length, and unknown tags.
- A valid event is accepted before and after its expiration timestamp; expiration is not an MLS/Marmot validity gate.
- Tampered ID/signature and routing/group-ID confusion are rejected before engine mutation.
- Direct and runtime ingest seams classify the same fixture identically.
- `just fast-ci` and targeted peeler/adapter/runtime tests pass.

### PR DESCRIPTION
**Pull request by erskingardner**
## Summary

- enforce the adopted exact kind-445 tag shape: one lowercase 32-byte `h`, optional syntactically valid NIP-40 `expiration`, and no other tags or extra tag elements
- validate the same shape at DTO mapping and peeler boundaries while treating expiration only as relay metadata, never a wall-clock validity rule
- add typed privacy-safe invalid-encoding, invalid-signature, and wrong-recipient outcomes; keep hostile terminal input from wedging transport drains
- cover direct, adapter, engine, and app runtime seams, including tampered IDs/signatures, routing-id confusion, and pre/post-expiration delivery

## Validation

- `cargo test -p cgka-traits`
- `cargo test -p transport-nostr-peeler`
- `cargo test -p transport-nostr-adapter`
- `cargo test -p cgka-engine`
- `cargo test -p marmot-app`
- `just fast-ci`

Fixes #1007

## Summary by CodeRabbit

* **Bug Fixes**
  * Reject invalid-signature and wrong-recipient messages earlier with correct typed stale outcomes, preventing further processing/telemetry.
  * Strengthened Nostr transport validation (including strict kind-445 tag shape and optional expiration handling).
  * Ensure repeated rejected messages reliably short-circuit to the “already seen” stale state.

* **Improvements**
  * Added privacy-safe typed inbound rejection categories and expanded audit-log peeler outcome kinds: `invalid_signature` and `wrong_recipient`.

* **Tests**
  * Expanded ingest and relay boundary tests to cover direct and snapshot-fallback typed terminal rejections, including Welcome-envelope wrong-recipient cases.

### CODE REVIEW

**Review by erskingardner — COMMENTED** [review]
# Code Review: PR #1081 — Enforce exact kind-445 ingress shape

## Summary

This PR tightens kind-445 (Marmot group message) tag validation to an exact adopted shape (one lowercase 32-byte `h`
tag, optional syntactically-valid NIP-40 `expiration`, nothing else), and splits the previously-generic
`PeelerError::Malformed` bucket into typed `InvalidSignature` / `WrongRecipient` outcomes that flow through to
privacy-safe, typed `TransportAdapterError` and forensic `PeelerOutcomeKind` variants. The mechanics are sound: the
exact-shape parser (`kind_445_transport_group_id`) is shared between DTO mapping and the peel-time routing check, the
new terminal-rejection classification is threaded consistently through direct peel and both snapshot-fallback seams
in `ingest.rs`, and the test coverage for the group-message path (tag-shape table, signature tampering, routing
confusion, snapshot-fallback boundary cases) is thorough. The one substantive gap is that the exact same
routing-mismatch fix applied to the group-message seam (`ensure_group_routing_matches`) was not mirrored onto the
structurally identical welcome-side check (`ensure_welcome_routing_matches`), which the repo's own engine
conventions call out as a bug class (mdk#707 "a guard that exists on one seam only is a bug").

## Issues

### Correctness/consistency: welcome routing-mismatch still classified `Malformed`, not `WrongRecipient`

**crates/transport-nostr-peeler/src/peeler.rs:479-498**

> ```rust
> fn ensure_welcome_routing_matches(
>     event: &NostrTransportEvent,
>     msg: &TransportMessage,
> ) -> Result<(), PeelerError> {
>     let event_recipient = event
>         .single_tag_value(RECIPIENT_TAG)
>         .map_err(to_peeler_error)
>         .and_then(|p| decode_hex_exact("recipient p tag", p, 32).map_err(to_peeler_error))?;
>     match &msg.envelope {
>         TransportEnvelope::Welcome { recipient } if recipient.as_slice() == event_recipient => {
>             Ok(())
>         }
>         TransportEnvelope::Welcome { .. } => Err(PeelerError::Malformed(
>             "event p tag does not match transport envelope".into(),
>         )),
>         TransportEnvelope::GroupMessage { .. } => Err(PeelerError::Malformed(
>             "welcome peeler received group envelope".into(),
>         )),
>     }
> }
> ```

Compare this to the sibling function this PR *did* update, `ensure_group_routing_matches` (same file, ~line 1111 in
the diff):

```rust
TransportEnvelope::GroupMessage { .. } => Err(PeelerError::WrongRecipient),
```

Both functions do exactly the same thing — check that a signed transport object's routing tag (`h` for group
messages, `p` for welcomes) agrees with the `TransportEnvelope` the transport layer attached to the message — and the
new `PeelerError::WrongRecipient` doc comment describes precisely this scenario:

> "The outer transport object names a routing id or recipient that does not match the transport envelope presented to
> the engine." (`crates/traits/src/error.rs`)

Yet only the group-message copy was migrated. The welcome copy still returns `Malformed`, and the existing test at
`crates/transport-nostr-peeler/src/peeler.rs:1081` (`welcome_peel_rejects_mismatched_p_tag_and_envelope`) still
asserts `PeelerError::Malformed(_)` — this PR left that assertion untouched, which is itself a signal the mismatch
was missed rather than intentional.

The gap compounds at the engine layer. `crates/cgka-engine/src/group_lifecycle.rs::terminal_welcome_error` was
updated to add `WrongRecipient` to its terminal-error match:

```rust
EngineError::Peeler(cgka_traits::error::PeelerError::InvalidSignature)
    | EngineError::Peeler(cgka_traits::error::PeelerError::WrongRecipient)
```

but `ingest_welcome` (`crates/cgka-engine/src/message_processor/ingest.rs:89-94`) collapses every
`terminal_welcome_error` match to a single `StaleReason::PeelFailed`, unlike the group-message seam which explicitly
special-cases `WrongRecipient` → `StaleReason::NotForThisClient` in three places (direct peel, and both
snapshot-fallback branches). So even after fixing `ensure_welcome_routing_matches`, the finer-grained stale-reason
and forensic (`PeelerOutcomeKind::WrongRecipient`) distinction this PR adds for group messages would still not
surface for the equivalent welcome case.

**Failure scenario:** a hostile or buggy relay/client delivers a welcome whose signed `p` tag names a different
pubkey than the `TransportEnvelope::Welcome{recipient}` the transport handed to the engine (transport/routing
confusion between two welcomes in flight) — the exact "routing-id confusion" class this PR is explicitly designed to
classify precisely. Under the current diff, this is still audited/reported as generic `malformed`/`PeelFailed`
instead of the new `wrong_recipient`/`NotForThisClient` classification, undermining the PR's own stated goal ("cover
direct, adapter, engine, and app runtime seams, including … routing-id confusion") for the welcome ingestion path.

Recommend: change `ensure_welcome_routing_matches`'s mismatch arm to `Err(PeelerError::WrongRecipient)`, update
`welcome_peel_rejects_mismatched_p_tag_and_envelope` accordingly, and have `ingest_welcome` discriminate
`WrongRecipient` to `StaleReason::NotForThisClient` (mirroring the group-message seam) rather than folding it into
`PeelFailed`.

## Suggestions

### Stale doc comment on `StaleReason::NotForThisClient`

**crates/traits/src/ingest.rs:39-40**

> ```rust
> /// A welcome addressed to a member other than ourselves.
> NotForThisClient,
> ```

This PR now also produces this reason for a *group message* whose `h` tag doesn't match its transport envelope
(routing-id confusion), not just for misaddressed welcomes. The doc comment wasn't updated and now undersells what
the variant covers. Worth a one-line addition mentioning the group-message routing-mismatch case, since a future
reader matching on this enum only has the doc comment to go on.

## What's Done Well

- `kind_445_transport_group_id` is the single source of truth for the exact tag shape, reused at both the DTO-mapping
  boundary (`to_transport_message`) and the peel-time routing check (`ensure_group_routing_matches`), which avoids
  exactly the kind of validation drift this repo's conventions warn about.
- The tag-shape table test (`kind_445_exact_tag_shape_table`) is a good compact way to cover the many malformed
  permutations (duplicate `h`, unknown tag, uppercase/short/non-hex encoding, duplicate/overflowing/negative
  `expiration`, tag order independence) without one test function per case.
- The snapshot-fallback terminal-rejection tests (`BoundaryRejectionPeeler` + `BoundaryRejectionPath`) correctly
  model the case where a direct peel returns `DecryptFailed`/`StaleEpoch` (deferring to snapshot fallback) and the
  *retained* snapshot then produces `InvalidSignature`/`WrongRecipient`, verifying the defense-in-depth branch that's
  documented as unreachable via the production Nostr peeler but still part of the public `TransportPeeler` contract.
- Signature/id verification order in `peel_group_message` is correct: `to_verified_nostr_event()` (id + signature) is
  checked before `ensure_group_routing_matches`, which runs before any decryption attempt — hostile input never
  reaches the decrypt step.
- The error-message changes are consistently privacy-safe: `PeelerError::InvalidSignature` /
  `PeelerError::WrongRecipient` and their `TransportAdapterError` counterparts carry fixed, non-attacker-controlled
  `Display` strings, and the adapter boundary (`map_inbound_event_error`) replaces a previous `Backend(format!(...))`
  path that echoed attacker-influenced detail into a generic error.
- `map_inbound_event_error` matches `NostrPeelerError` exhaustively (no wildcard arm), so a future variant addition
  will fail to compile here instead of silently defaulting — good insurance against the exact class of gap flagged
  above.

File: crates/marmot-forensics/schema/audit-log-event.v2.schema.json [hunk 1]
```diff
@@ -693,7 +693,15 @@
...
-      "enum": ["success", "decrypt_failed", "stale_epoch", "malformed", "other"]
+      "enum": [
+        "success",
+        "decrypt_failed",
+        "stale_epoch",
+        "malformed",
+        "invalid_signature",
+        "wrong_recipient",
+        "other"
+      ]
```
[the remark this answers is not in the document]
Comment (erskingardner) [reply crates/marmot-forensics/schema/audit-log-event.v2.schema.json:704]: Verified against the current ingest mapping: `PeelerOutcomeKind` emits `Other` for `MissingContext`, `WrapFailed`, and `Backend`; those exact strings appear only in the optional `error_kind` detail field, whose schema is intentionally an unrestricted string. The typed outcome enum and v2 schema are already aligned, so adding these values to `peelerOutcomeKind` would misrepresent the Rust catalog.
